core/test/base/range_accessors.cpp

```diff
@@ -121,4 +139,569 @@ TEST_F(RowMajorAccessor, CanAssignSubranges)
+        std::tuple<int, int, int> ignore = std::tuple<int, int, int>(99, 99,
+                                                                     99))
+    {
+        // Test for equality is fine here since they should not be modified
+        // clang-format off
+        if (ignore != t(0, 0, 0)) { EXPECT_EQ(a(0, 0, 0), 10.);   }
+        if (ignore != t(0, 0, 1)) { EXPECT_EQ(a(0, 0, 1), 22.);   }
+        if (ignore != t(0, 1, 0)) { EXPECT_EQ(a(0, 1, 0), -12.);  }
+        if (ignore != t(0, 1, 1)) { EXPECT_EQ(a(0, 1, 1), 26.);   }
+        if (ignore != t(0, 2, 0)) { EXPECT_EQ(a(0, 2, 0), 14.);   }
+        if (ignore != t(0, 2, 1)) { EXPECT_EQ(a(0, 2, 1), -230.); }
+        if (ignore != t(0, 3, 0)) { EXPECT_EQ(a(0, 3, 0), 6.);    }
+        if (ignore != t(0, 3, 1)) { EXPECT_EQ(a(0, 3, 1), 154.);  }
+        // clang-format on
+    }
+};
+
+
+TEST_F(ScaledReducedStorage3d, CorrectLengths)
+{
+    EXPECT_EQ(r.length(0), size[0]);
+    EXPECT_EQ(r.length(1), size[1]);
+    EXPECT_EQ(r.length(2), size[2]);
+    EXPECT_EQ(r.length(3), 1);
+    EXPECT_EQ(r->get_size(), size);
+}
+
+
+TEST_F(ScaledReducedStorage3d, CorrectStride)
+{
+    EXPECT_EQ(r->get_stride()[0], size[1] * size[2]);
+    EXPECT_EQ(r->get_stride().at(0), size[1] * size[2]);
+    EXPECT_EQ(r->get_stride()[1], size[2]);
+    EXPECT_EQ(r->get_stride().at(1), size[2]);
+}
+
+
+TEST_F(ScaledReducedStorage3d, CorrectStorage)
+{
+    EXPECT_EQ(r->get_storage(), data);
+    EXPECT_EQ(r->get_const_storage(), data);
+}
+
+
+TEST_F(ScaledReducedStorage3d, CorrectScale)
+{
+    EXPECT_EQ(r->get_scale(), scale);
+    EXPECT_EQ(r->get_const_scale(), scale);
+}
+
+
+TEST_F(ScaledReducedStorage3d, CanReadData)
+{
+    check_accessor_correctness(r);
+    check_accessor_correctness(cr);
+}
+
+
+TEST_F(ScaledReducedStorage3d, CopyFrom)
+{
+    st_type data2[data_elements];
+    ar_type scale2[scale_elements];
+    reduced_storage cpy(data2, scale2, size);
+
+    // Do not use this in regular code since the implementation is slow
+    cpy = r;
+
+    check_accessor_correctness(cpy);
+}
+
+
+TEST_F(ScaledReducedStorage3d, CanImplicitlyConvertToConst)
+{
+    const_reduced_storage const_rs = r->to_const();
+    const_reduced_storage const_rs2 = cr;
+
+    check_accessor_correctness(const_rs);
+    check_accessor_correctness(const_rs2);
+}
+
+
+TEST_F(ScaledReducedStorage3d, ToConstWorking)
+{
+    auto cr2 = r->to_const();
+
+    static_assert(std::is_same<decltype(cr2), const_reduced_storage>::value,
+                  "Types must be equal!");
+    check_accessor_correctness(cr2);
+}
+
+
+TEST_F(ScaledReducedStorage3d, CanRead)
+{
+    check_accessor_correctness(cr);
+    check_accessor_correctness(r);
+}
+
+TEST_F(ScaledReducedStorage3d, Subrange)
+{
+    auto subr = cr(0, gko::span{0, 2}, 1);
+
+    EXPECT_EQ(subr(0, 0, 0), 22.);
+    EXPECT_EQ(subr(0, 1, 0), 26.);
+}
+
+
+TEST_F(ScaledReducedStorage3d, CanWriteScale)
+{
+    r->write_scale(0, 0, 10.);
+
+    EXPECT_EQ(r(0, 0, 0), 100.);
```

Comment:
Maybe I am misunderstanding something here but I thought `write_scale` just writes a scalar value at a certain index [code doc](https://github.com/ginkgo-project/ginkgo/blob/add_new_accessor/include/ginkgo/core/base/range_accessors.hpp#L807). Here it seems to be scaling the existing the value by the passed in value at the index ?

Reply:
`write_scale` (now renamed to `write_scalar`) writes the given value to the scalar location. That means that the storage is interpreted differently, since all reads of the accessor (and not the scalar function) basically do `scalar * stored_value`, and all writes `stored_value = given_value / scalar`. That means a change in scale changes the way the memory is interpreted (that uses that specific scalar).